text/0045-api.md

```diff
@@ -0,0 +1,191 @@
+        }
+        JsonValue::Object(json_obj)
+    }
+}
+```
+
+Notable is the use of `HashItem` or `TransactionRefItem` in the conversions. These wrapper types are used to be conform with Rust's orphan rules. Since multiple services use for example the `Hash` type within parameters/response, conversions should be reusable.
+For this reason, `TryFrom/From`s should be defined also for inner types (e.g. `Hash`) of a request/response.
+```
+pub struct HashItem(pub Hash);
+
+impl TryFrom<&str> for HashItem {
+    type Error = &'static str;
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        ...
+    }
+}
+
+impl From<&HashItem> for String {
+    fn from(value: &HashItem) -> Self {
+        ...
+    }
+}
+
+impl From<&HashItem> for JsonValue {
+    fn from(value: &HashItem) -> Self {
+        JsonValue::String(String::from(value))
+    }
+}
+```
+
+Some types require more conversion work than others. For consistency - if desired - `TryFrom`/`From`s could be defined for all inner types of a request/response. An example that requires less conversion work, but still might get its `From` implementation:
+```
+pub struct MilestoneIndexItem(pub MilestoneIndex);
+
+impl From<&MilestoneIndexItem> for JsonValue {
+    fn from(value: &MilestoneIndexItem) -> Self {
+        JsonValue::from(*value.0)
+    }
+}
+```
+
+## WebService trait
+
+The `WebService` trait specifies which endpoints a Web service should implement. This also means, it doesn't need to expose all functions of the `Service` trait.
+
+```
+#[async_trait]
+pub trait WebService {
+    type Input;
+    type Output;
+
+    async fn node_info() -> Self::Output;
+    async fn transactions_by_bundle(input: Self::Input) -> Self::Output;
+    async fn transaction_by_hash(input: Self::Input) -> Self::Output;
+    async fn transactions_by_hashes(input: Self::Input) -> Self::Output;
+}
+```
+
+The associated types are remarkable. Since some Web services might require a certain function signature, this flexibility is required.
+In the case of `warp` the `WebService` is implemented as follows:
+```
+#[async_trait]
+impl WebService for Rest {
+    type Input = JsonValue;
+    type Output = Result<warp::reply::Json, warp::Rejection>;
+
+    async fn node_info() -> Self::Output {
+        Ok(warp::reply::json(&JsonValue::from(ServiceImpl::node_info())))
+    }
+    ...
+}
+```
+
+These implemented functions then can be easily injected into the Web service.
+In the case of `warp` it would look like this:
+```
+warp::get()
+    .and(warp::path("v1"))
+    .and(warp::path("node-info"))
+    .and(warp::path::end())
+    .and_then(routes::Rest::node_info);
+```
+
+# Drawbacks
+
+No drawbacks known to the author.
+
+# Rationale and alternatives
+
+No viable alternatives known to the author.
+
+# Unresolved questions
+
+- Should function params be created with a `new` constructor instead? And eventually return an error already there when illegal parameters where passed, instead in the service?
```

Comment:
I do feel like a builder pattern is more idiomatic in Rust. But that's personal taste if the interface is not going to be used by others (that is only bee devs and contributors need to build the layer by these traits.)

Reply:
Good idea, I guess builder pattern would be nice here for parameters/response